Automattic/wp-calypso
# Webpack monitor: No console interception 😁

#### Changes proposed in this Pull Request

* Drop console interception from WebpackBuildMonitor
* Implement WebpackBuildMonitor updates with
  - EventStream (webpack HMR sends updates via EventStream)
  - Websockets (Our non-Webpack CSS hot updates support websockets)
* Rewrite in TypeScript as a Function component

The most important change here is that we do not need to wrap/intercept the global console. For a long time, this has made stack traces for `console.warn` inaccurate, but this should restore the correct trace.

##### Before

![Screen Shot 2019-08-28 at 17 19 53](https://user-images.githubusercontent.com/841763/63869391-5b091c80-c9b8-11e9-8bd9-da67d81054ad.png)


##### After

![Screen Shot 2019-08-28 at 17 20 39](https://user-images.githubusercontent.com/841763/63869402-5f353a00-c9b8-11e9-99e1-f369a1f02d15.png)


#### Issues

The `Needs Reload` state may not be perfectly reliable, I'm not entirely sure. We don't seem to have any hot reloading so this is probably a fine tradeoff.

#### Testing instructions

* Spin up this PR in development.
* You can use `localStorage.debug = 'calypso:webpack-build-monitor'` to get debug output
* Make changes to CSS ([there aren't many](https://github.com/Automattic/wp-calypso/blob/39b09573180a669aa705bd93484a7d836c71eed0/assets/stylesheets/_components.scss)) and/or Webpack-controlled files (most JS/CSS in the application).
* The build monitor should update accordingly responding to changes.
* Add `console.warn` somewhere and verify that the trace in the console comes from the right place 🎉
